from concurrent import futures
import grpc

import chat_pb2
import chat_pb2_grpc

# Armazena o hist칩rico de mensagens em mem칩ria
messages = []


class ChatHistoryService(chat_pb2_grpc.ChatHistoryServicer):
    """
    Implementa칞칚o do servi칞o definido no arquivo chat.proto.

    Respons치vel por:
    - Salvar mensagens recebidas
    - Retornar o hist칩rico completo de mensagens
    """

    def SaveMessage(self, request, context):
        """
        RPC respons치vel por salvar uma nova mensagem no hist칩rico.

        Par칙metros:
        - request.user: nome do usu치rio que enviou a mensagem
        - request.text: conte칰do da mensagem

        Retorno:
        - Empty (confirma apenas que a opera칞칚o foi conclu칤da)
        """

        messages.append({
            "user": request.user,
            "text": request.text
        })

        return chat_pb2.Empty()

    def ListMessages(self, request, context):
        """
        RPC respons치vel por retornar todas as mensagens salvas.

        Retorna:
        - MessageList contendo uma lista de mensagens
        """

        response = chat_pb2.MessageList()

        for msg in messages:
            response.messages.add(
                user=msg["user"],
                text=msg["text"]
            )

        return response


def serve():
    """
    Inicializa e executa o servidor gRPC.

    - Cria um pool de threads para atender m칰ltiplas requisi칞칫es
    - Registra o servi칞o ChatHistoryService
    - Escuta na porta 8000
    """

    server = grpc.server(
        futures.ThreadPoolExecutor(max_workers=10)
    )

    chat_pb2_grpc.add_ChatHistoryServicer_to_server(
        ChatHistoryService(),
        server
    )

    # Porta onde o servi칞o gRPC ficar치 dispon칤vel
    server.add_insecure_port("[::]:8000")

    server.start()
    print("游 Servidor gRPC ativo\n")

    # Mant칠m o servidor em execu칞칚o
    server.wait_for_termination()


if __name__ == "__main__":
    serve()
